### PR TITLE
Update BasicAuthenticate To Support Different Encodings Than ISO-8859-1

### PR DESCRIPTION
**Pull request by Manuel1948**
The problem is that basic authentication credentials are encoded in ISO-8859-1, but if you use an other encoding f.e. UTF-8 and the Auth component compares a username or password with special chars it comes to errors.

So I added a check if there is set another valid encoding than ISO-8859-1 and if so, convert it to the actual encoding.

### CODE REVIEW

**Comment by Manuel1948** [comment]
Issue #10359

**Comment by cleptric** [comment]
Could please add a test case for that.

File: src/Auth/BasicAuthenticate.php [hunk 1]
```diff
@@ -75,6 +76,11 @@ public function getUser(ServerRequest $request)
     {
         $username = $request->env('PHP_AUTH_USER');
         $pass = $request->env('PHP_AUTH_PW');
+        $encoding = Configure::read('App.encoding');
+        if ($encoding && in_array($encoding, mb_list_encodings()) && $encoding != 'ISO-8859-1') {
```
Comment (dereuromark) [inline src/Auth/BasicAuthenticate.php]: Please use !==

File: src/Auth/BasicAuthenticate.php [hunk 2]
```diff
@@ -75,6 +76,11 @@ public function getUser(ServerRequest $request)
...
+            $username = mb_convert_encoding($username, $encoding, 'ISO-8859-1');
+            $pass = mb_convert_encoding($pass, $encoding, 'ISO-8859-1');
```
Comment (dereuromark) [inline src/Auth/BasicAuthenticate.php]: Ideally that string is a constant in that class, so we don't have to use a magic string 3 times (with the chance of typos).

**Comment by cleptric** [comment]
I'm also not sure, if they are always ` ISO-8859-1` encoded. Isn't that depending of our php.ini settings?

File: src/Auth/BasicAuthenticate.php (hunk 2 above)
Comment (Manuel1948) [inline src/Auth/BasicAuthenticate.php]: thx

File: src/Auth/BasicAuthenticate.php (hunk 1 above)
Comment (Manuel1948) [inline src/Auth/BasicAuthenticate.php]: thx

**Comment by dereuromark** [comment]
Indeed, this sounds like we should get the input encoding from a more dynamic source rather than a hardcoded string.

**Comment by cleptric** [comment]
Which then leads me to the questions if this can't be fixed by just updating your php.ini?

**Comment by Manuel1948** [comment]
Dear cleptric,

can you suggest me a change in php.ini because my default charset is UTF-8 but this not the solution :/

**Comment by Manuel1948** [comment]
http://toscho.de/2009/authentifizierung-php-utf-8/ 

I edited the commits again based on this page.

File: src/Auth/BasicAuthenticate.php [hunk 3]
```diff
@@ -75,6 +77,11 @@ public function getUser(ServerRequest $request)
     {
         $username = $request->env('PHP_AUTH_USER');
         $pass = $request->env('PHP_AUTH_PW');
+        $encoding = Configure::read('App.encoding');
+        if ($encoding && in_array($encoding, mb_list_encodings())) {
+            $username = mb_convert_encoding($username, $encoding, self::ENCODING_SOURCES);
```
Comment (dereuromark) [inline src/Auth/BasicAuthenticate.php]: Late static binding (static::) helps to allow extensions to work with their own const.

Comment (Manuel1948) [inline src/Auth/BasicAuthenticate.php]: I'm a bit confused now, should I correct the :: to -> or is it okay?

Comment (ADmad) [inline src/Auth/BasicAuthenticate.php]: You need to use `::static` instead of `::self`

Comment (Manuel1948) [inline src/Auth/BasicAuthenticate.php]: Thank you, I'm not that expert as you are

**Review by markstory — CHANGES_REQUESTED** [review]
A test case is required to merge this. It will regress to easily in the future without a test case.

**Comment by Manuel1948** [comment]
Hope someone can write a test case, because I can't :/

**Comment by Manuel1948** [comment]
This problem isn't currently reproducable anymore, looks like something has changed this behavior.

**Pull request closed without merging**